helix-view/src/clipboard/config.rs

```diff
@@ -0,0 +1,248 @@
+use std::ffi::OsStr;
+
+use log::warn;
+use serde::{Deserialize, Serialize};
+
+use super::{provider, ClipboardProvider};
+
+#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
+#[serde(from = "CustomClipboardConfig")]
+pub struct ClipboardConfig {
+    providers: HashMap<String, ClipboardProviderSpec>,
+    order: Vec<String>,
+}
+
+#[derive(Deserialize)]
+struct CustomClipboardConfig {
+    #[serde(rename = "provider")]
+    providers: HashMap<String, ClipboardProviderSpec>,
+    order: Vec<String>,
+}
+
+impl From<CustomClipboardConfig> for ClipboardConfig {
+    fn from(cfg: CustomClipboardConfig) -> Self {
+        let mut providers = Self::default().providers;
+        providers.extend(cfg.providers);
+        let order = cfg.order;
+        Self { providers, order }
+    }
+}
+
+#[derive(Debug, Default, Clone, PartialEq, Deserialize, Serialize)]
+#[serde(rename_all = "kebab-case")]
+struct ClipboardProviderSpec {
+    copy: Vec<String>,
+    paste: Vec<String>,
+    primary_copy: Option<Vec<String>>,
+    primary_paste: Option<Vec<String>>,
+    #[serde(default)]
+    env_vars: Vec<String>,
+    #[serde(default)]
+    test_commands: Vec<Vec<String>>,
+}
+
+macro_rules! command_provider_spec {
+    (
+        copy: $($get:literal),+ ;
+        paste: $($set:literal),+ ;
+        $(
+            primary_copy: $($pget:literal),+ ;
+            primary_paste: $($pset:literal),+ ;
+        )?
+        $( env: $($env:literal),+ ; )?
+        $( test: $($test:literal),+ ; )*
+    ) => {
+        #[allow(clippy::needless_update)]
+        ClipboardProviderSpec {
+            copy: vec![$($get.to_owned()),+],
+            paste: vec![$($set.to_owned()),+],
+            $(
+                primary_copy: Some(vec![$($pget.to_owned()),+]),
+                primary_paste: Some(vec![$($pset.to_owned()),+]),
+            )?
+            $( env_vars: vec![$($env.to_owned()),+], )?
+            test_commands: vec![$(vec![$($test.to_owned()),+]),*],
+            ..Default::default()
+        }
+    };
+}
+
+macro_rules! clipboard_config {
+    ($(
+        $(#[$attr:meta])*
+        $name:literal {
+            $($body:tt)*
+        }
+    )*) => {
+        ClipboardConfig {
+            providers: HashMap::from([$(
+                $(#[$attr])*
+                ($name.to_owned(), command_provider_spec! { $($body)* })
+            ),*]),
+            order: vec![$(
+                $(#[$attr])*
+                $name.to_owned()
+            ),*],
+        }
+    };
+}
+
+impl Default for ClipboardConfig {
+    fn default() -> Self {
+        let mut cfg = clipboard_config! {
+            #[cfg(unix)]
+            "wl-clipboard" {
+                copy: "wl-copy", "--type", "text/plain";
+                paste: "wl-paste", "--no-newline";
+                primary_copy: "wl-copy", "-p", "--type", "text/plain";
+                primary_paste: "wl-paste", "-p", "--no-newline";
+                env: "WAYLAND_DISPLAY";
+            }
```

Comment:
Do we really need a macro for this?

Reply:
I mean, there was a macro before, and it didn't reduce boilerplate by nearly as much as these two macros.
This approach handles:
- Converting every argument into a `String` so that the presets can coexist with deserialized config data.
- Building a map of providers, for by-name lookup.
- Building a list based on order of appearance, for specifying precedence.
- Implicit defaults for `env`, `primary`, and `test`, which are each relatively rare among the presets.